### PR TITLE
Enable declare v3

### PR DESCRIPTION
**Pull request by ksew1**
**Stack**:
- #2310
- #2309
- #2308 ⬅
- #2307

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: crates/sncast/src/helpers/fee.rs [hunk 1]
```diff
@@ -151,3 +151,38 @@ pub enum FeeSettings {
         max_gas_unit_price: Option<u128>,
     },
 }
+
+#[macro_export]
+macro_rules! impl_from_version_for_fee_token {
```
Comment (Arcticae) [inline crates/sncast/src/helpers/fee.rs:156]: I really like the solution, although i wonder: 
Couldn't this be solved with a trait? 
Something like `PayableTransaction`, which would expose the mapping of version -> token 
And also would define a method for formatting the error

WDYT?

**Review by Arcticae — APPROVED** [review]
lgtm overall, one nitpick/q

File: crates/sncast/src/helpers/fee.rs (hunk 1 above)
Comment (ksew1) [reply crates/sncast/src/helpers/fee.rs:156]: I like this idea but I want to be sure we're on the same page:

I am thinking to have a trait like this:

```rust
trait PayableTransaction {
    fn error_message(&self, token: &str, version: &str) -> String;
    fn validate(&self) -> Result<()>;
    fn version_from_token(&self) -> Option<FeeToken>;
}
```

and implement it via a macro:

```rust
impl_payable_transaction!(Invoke, token_not_supported_for_invoke,
    InvokeVersion::V1 => FeeToken::Eth,
    InvokeVersion::V3 => FeeToken::Strk
);
```

Comment (Arcticae) [reply crates/sncast/src/helpers/fee.rs:156]: Yep i am not against implementing it via a macro
Having the clear interface without `cargo expand` is what i'm after

**Comment by ksew1** [comment]
✓ Commit merged in pull request [#2310](https://github.com/foundry-rs/starknet-foundry/pull/2310)

**Pull request closed without merging**